### PR TITLE
Code selection by column on DIFF screen

### LINKED ISSUE #3089: Code selection "by column" on diff screen
It can be useful to copy/paste the code while comparing branches (diff view) so that to copy easily some portions of code from branch to branch.

Unfortunately, in the diff screen of abapGit, whatever it's in the split or unified view, the selection of code selects the whole lines, including the row numbers and marks, and if it's in the split view, the code of the other side is also included.

As-is (want to copy code from the right side, but unfortunately the code from the left side is also selected):

image

To-be:

image

For information, the principle how to select and copy code of an HTML table column is explained here: https://stackoverflow.com/questions/6619805/select-text-in-a-column-of-an-html-table.

I have prototyped the solution and it works well (for me). I will propose a pull request if you agree with this feature.

### PR DESCRIPTION
**Pull request by sandraros**
Fixes #3089

Two ways to select a text:

1. Position the mouse at the start of the text, press mouse left button, keep pressed, drag to the end of the text, release the button

2. Position the mouse at the start of the text, press mouse left button, release the button, position the mouse to the end of the text, press shift key + mouse left button.

Two ways to copy the selected text:

1. Press Ctrl + C or Ctrl + Ins

2. Mouse right button on the text to display the context menu and choose Copy

SPLIT VIEW (code at the left is the LOCAL code and code at the right is the REMOTE code):
- Text can be selected only at the left side, or only at the right side.
- The dummy empty lines which show the place of lines which exist only at the other side are not copied.

UNIFIED VIEW (LOCAL and REMOTE are mixed in the same column, "green plus" lines correspond to lines only in the LOCAL code, "red minus" lines correspond to lines only in the REMOTE code.
- Only the lines from the REMOTE code are copied

PATCH VIEW (split view + one extra column on the left)
- Same features as the SPLIT VIEW

### CODE REVIEW

File: src/ui/zabapgit_js_common.w3mi.data.js [hunk 1]
```diff
@@ -1837,3 +1840,103 @@ function enumerateJumpAllFiles() {
...
+    nodes = doc.querySelectorAll("tr"),
+    text = "";
+
+  if (nodes.length === 0) {
+    text = doc.textContent;
+  } else {
+    // special processing for TD tag which sometimes contains newline
+    // (/src/ui/zabapgit_js_common.w3mi.data.js) so don't add newline in that case.
+    var newline = "";
+    [].forEach.call(nodes, function(tr, i) {
```
Comment (FreHu) [inline src/ui/zabapgit_js_common.w3mi.data.js]: I think you can just do `nodes.forEach` here.

Comment (sandraros) [reply src/ui/zabapgit_js_common.w3mi.data.js]: Unfortunately, it works in Chrome but it doesn't in SAP GUI.

**Comment by sbcgua** [comment]
I have some suggestions regarding JS code style. It contains global variables (`selectedColumnIdx`, `lineNumColumnIdx`) and also function names are too unspecific (`addMousedownEventListener`, ...). I's suggest wrapping it into a class. This would remove globals, introduce a clear namespace and reduce JS initiation code in abap (to `new ...`). 
Also numerical values are not that clear e.g. `5 + patchColumns` - why 5 ? Maybe add a named variable for this e.g. `remoteStartColumn` or similar ?

**Comment by sandraros** [comment]
@sbcgua I used `prototype` because `class` is invalid in the old javascript of the SAP GUI HTML control, and did the changes as suggested.

File: src/ui/zabapgit_js_common.w3mi.data.js [hunk 2]
```diff
@@ -1841,3 +1841,151 @@ function enumerateJumpAllFiles() {
...
+  this.lineNumColumnIdx = -1;
+}
+
+function DiffColumnSelection_addMousedownEventListener() {
+  document.addEventListener("mousedown",
+    //https://stackoverflow.com/questions/2749244/javascript-setinterval-and-this-solution
+    (function(self) {
+      return function(e) {
+        self.mousedownEventListener(e);
+      };})(this));
```
Comment (sbcgua) [inline src/ui/zabapgit_js_common.w3mi.data.js]: there is a `bind` polyfill at the top, so the code can simplified like this

```js
function DiffColumnSelection_addMousedownEventListener() {
  //https://stackoverflow.com/questions/2749244/javascript-setinterval-and-this-solution
  document.addEventListener("mousedown", this.mousedownEventListener.bind(this));
}
```

File: src/ui/zabapgit_js_common.w3mi.data.js [hunk 3]
```diff
@@ -1841,3 +1841,151 @@ function enumerateJumpAllFiles() {
...
+    this.lineNumColumnIdx = -1;
+  }
+}
+
+function DiffColumnSelection_addCopyEventListener() {
+  document.addEventListener("copy",
+    (function(self) {
+      return function(e) {
+        self.copyEventListener(e);
+      };})(this));
```
Comment (sbcgua) [inline src/ui/zabapgit_js_common.w3mi.data.js]: same here (`bind`) ?

File: src/ui/zabapgit_js_common.w3mi.data.js [hunk 4]
```diff
@@ -1841,3 +1841,151 @@ function enumerateJumpAllFiles() {
...
+
+function DiffColumnSelection_Initialize() {
+  DiffColumnSelection.prototype.addMousedownEventListener = DiffColumnSelection_addMousedownEventListener;
+  DiffColumnSelection.prototype.mousedownEventListener = DiffColumnSelection_mousedownEventListener;
+  DiffColumnSelection.prototype.addCopyEventListener = DiffColumnSelection_addCopyEventListener;
+  DiffColumnSelection.prototype.copyEventListener = DiffColumnSelection_copyEventListener;
+  DiffColumnSelection.prototype.getSelectedText = DiffColumnSelection_getSelectedText;
+  var columnSelection = new DiffColumnSelection();
+  return columnSelection;
+}
```
Comment (sbcgua) [inline src/ui/zabapgit_js_common.w3mi.data.js]: what about:

```js
function DiffColumnSelection() {
    this.selectedColumnIdx = -1;
    // construction stuff
}
DiffColumnSelection.prototype.addMousedownEventListener = function() {
    ...
}
```

File: src/ui/zabapgit_js_common.w3mi.data.js (hunk 2 above)
Comment (sandraros) [reply src/ui/zabapgit_js_common.w3mi.data.js]: all changes done as requested

Comment (sandraros) [reply src/ui/zabapgit_js_common.w3mi.data.js]: done

**Comment by sbcgua** [comment]
@sandraros the code looks fine, thanks ! I'll make a test later today and confirm

**Comment by sbcgua** [comment]
@sandraros one small question, you are using `== undefined` and `!= undefined`, is it intentional ? The general best practice is to use `===` which also compares types. E.g. `null == undefined` is true. I guess in this code it is OK ... but just asking

**Comment by sandraros** [comment]
@sbcgua Not intentional. But I tried to change it and it does an error "SCRIPT5007: Unable to get property 'XXXXXX' of undefined or null reference" and it breaks the scrollbar because click event is outside td, same as copy outside td. So, it doesn't work (maybe it's because of bad code around but I'm not expert) and is intentional now.
I propose a new commit because of some unused lines I forgot to remove.

**Comment by sbcgua** [comment]
works nicely :)

**Comment by sandraros** [comment]
@sbcgua Now that it's almost done, thank you, I appreciated your constructive comments and learned a lot, I hope to do it in less steps next time :)

**Pull request merged**